### DISCUSSION TITLE
How to configure Keepalived as L4 load balancer via Direct Routing (DR) method?

### DISCUSSION THREAD

**Discussion by moaali** [acassen/keepalived#2314] [Q&A]
Trying to achieve L4 load balancing via Keepalived in front of HAProxy that will act as L7 load balancer. Both Keepalived and HAProxy are on separate machines. I managed to get everything in the below image working but when I try to send a request to the public virtual IP (i.e. 115.101.1.17), **the connection always times out**.

As per my understanding from the documentation that states:

> In order for the real servers to directly respond to the public users’ requests, each real server must use the VIP as its source address when sending replies.

I tried to reset the source IP using SNAT on the HAProxy machine but still the same thing happens.

**Rules used for iptables on real servers (HAProxy machines)**

```
iptables \
  -A POSTROUTING \
  -t nat \
  -p tcp \
  --dport 80 \
  -j SNAT \
  --to-source 115.101.1.17
```

**keepalived.conf**

```
vrrp_instance VI_1 {
  state       MASTER     # [1]
  interface   eth0       # [2]
  advert_int  1          # [3]
  priority    100        # [4]

  virtual_router_id  92  # [5]

  # Authentication for VRRP messages
  authentication {
    auth_type  PASS
    auth_pass  pass123
  }

  virtual_ipaddress {
    115.101.1.17 dev eth0  # [6]
  }
}

virtual_server 115.101.1.17 80 {
  lb_algo   rr             # [1]
  lb_kind   DR             # [2]
  protocol  TCP

  delay_loop           10  # [3]
  persistence_timeout  60  # [4]

  # Backend Server (HAProxy LB-01)
  real_server 10.0.1.2 80 {
    weight 100

    TCP_CHECK {
      connect_timeout 5
      nb_get_retry 3
      delay_before_retry 2
    }
  }

  # Backend Server (HAProxy LB-02)
  real_server 10.0.1.3 80 {
    weight 100

    TCP_CHECK {
      connect_timeout 5
      nb_get_retry 3
      delay_before_retry 2
    }
  }
}
```

**sysctl.conf**

```
net.ipv4.ip_nonlocal_bind = 1
net.ipv4.ip_forward = 1
```

Any help will be appreciated.

[![image of keepalived and haprxoy setup][1]][1]

  [1]: https://i.stack.imgur.com/CwCLh.png

**Comment by nser77** [comment]
Hi, in your case, ```connection time out``` is an expected behavior; the main reason is that ```keepalived``` and ```haproxy``` servers are on different subnets (```10.0.0.0/24``` and ```10.0.1.0/24```) and loadbalancig direct routing works at MAC Address level; based on your scenario, ```haproxy``` servers will never recive traffic beacuse ```keepalived``` is not able to learn the (real) MAC Address of those real servers - they are separated by a L3 device. 

We can confirm that from your ```virtual_server``` configuration block:

```
[...]

  real_server 10.0.1.2 80 {

[...]

  real_server 10.0.1.3 80 {

[...]
```

Besides this, I think it is not enough and there might be more suggestions as well:

1. If ```lvs_method``` (or ```lb_kind```) is set to ```DR```, ```keepalived``` will obviously forward packets as a transparent proxy (see above) so the source and destination IP Addresses will not be changed; in that case, your real server ```haproxy``` would receive an IP packet like this: ```23.4.1.2:44732 -> 10.0.0.2:443```; you must accept that traffic on all real servers to create the socket; RedHat has an interesting article [1].

2. You might need to specify/clarify which device acts as L3.

3. You may want to traverse ```keepalived``` [trasparent] proxy with SSL.

4. Checks from ```keepalived``` for real servers should be on port 443 and not port 80.

5. ```net.ipv4.ip_forward=1``` is not needed in this case, leave it disabled.

----

[1] _3.2. Load Balancer Using Direct Routing_: https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/load_balancer_administration/s1-lvs-direct-vsa#doc-wrapper

**Comment by pqarmitage** [comment]
@moaali Do you still need assistance with this, or have you resolved your problems?

As a starting point I would look at the Linux Virtual Server web site, e.g. http://linuxvirtualserver.org/VS-DRouting.html for direct routing.

When you use Direct Routing, IPVS forwards the incoming packets without modifying the IP layer, so a TCP packet destination port 80 received from 1.2.3.4 to 115.101.1.17 will be forwarded to the chosen real server using the same IP addresses. It uses the configured IP address of the real server in order to set the destination MAC address for the forwarded packet. This means two things:

1. The system on which keepalived is running must be able to find the MAC address associated with IP addresses 10.0.1.2, 10.0.1.3 and 10.0.1.4, This mean that the system on which keepalived is running probably needs an IP address on 10.0.1.0/24, such as 10.0.1.1/24 which does not appear to be being used.

2. When a real server receives a packet forwarded by IPVS, the destination MAC address will be 115.101.1.17. This means that 115.101.1.17 must be configured as a local address on each of the real servers (I tend to create an additional loopback interface and configure the address on that) or alternatively use iptables/nftables DNAT for 115.101.1.17 to a local address. Either way, since HAProxy will reply on the same socket as it received the packet on, return packets will have the source IP address 115.101.1.17, and so an iptables SNAT entry is not required (BTW the iptables entry you show above should use `--sport` rather than `--dport` (you can check whether the rule is being used by looking at the counters in the `iptables -t nat -nvL POSTROUTING` output)).

The final issue that occurs to me is the return path for packets back to the internet. The example shown on the Linux Virtual Server web page referred to above shows return packets not going back via the keepalived server, whereas it is not clear what the return path is in your configuration. One potential issue to check, if the return path is via the keepalived server, is whether it is happy forwarding packets that it receives with the source address of the packet matching one of its own IP addresses (i.e. 115.101.1.17).

**Comment by tomachinz** [comment]
Hello pqarmitage,
do you recommend lb_kind DR or lb_kind NAT in general?
I would assume DR, interested to hear your response. Perhaps NAT is better if the real servers do not have the virtual IP address configured as an interface; and perhaps DR is better if you DO have the virtual IP configured.
After reading your message, I've decided to switch to NAT for my setup until such time as I know how to configure another (multi-homed?) IP on the real servers, which raises the question:

How best should I configure my real servers with this new loopback? Similar to how you said: "create an additional loopback interface and configure the address on that" it would seem to me that this interface will need to have gateway is that true? Unlike the loopback.

In my case, everything is in same subnet, my VIP is 10.0.0.100 and one of the real servers is 10.0.0.122 (ns2) currently this is in state BACKUP and MASTER is a different machine ns1. In 'ip a' I see gateway lines as follows, my loopback:
`    inet 127.0.0.1/8 scope host lo`
backup server ns2 real interface eth0@if12:
`    inet 10.0.0.122/24 brd 10.0.0.255 scope global dynamic eth0`
correct me if I wrong but once I've configured an additional interface for the virtual IP on a server, the gateway shown with 'ip a' should probably appear similar to:
`    inet 10.0.0.100/24 brd 10.0.0.255 scope host lo`
surely my guess is wrong? I replaced the scope part do I also need to remove broadcast? 
how do I configure this VIP interface please? reading that page my guess is:

The real server ns2, with IP forwarding enabled, seems to only require two lines:

```
    ifconfig lo:0 10.0.0.100 netmask 255.255.255.255 broadcast 10.0.0.100 up
    route add -host 10.0.0.100 dev lo:0
```

is broadcast address same as VIP OK (above)? or should I correct to say 10.0.0.255?
I think I can skip these lines:

```
    ifconfig eth0 10.0.0.122 netmask 255.255.255.0 broadcast 10.0.0.255 up
    route add -net 10.0.0.0 netmask 255.255.255.0 dev eth0
```

because I already see real server has IP .122 given by DHCP and broadcast .255 as above. WOW I THINK ITS WORKING.

```
-bash: ifconfig: command not found
root@ns2:/etc/keepalived# uname -a
Linux ns2 6.8.4-2-pve #1 SMP PREEMPT_DYNAMIC PMX 6.8.4-2 (2024-04-10T17:36Z) x86_64 GNU/Linux
 apt install net-tools  
    ifconfig lo:0 10.0.0.100 netmask 255.255.255.255 broadcast 10.0.0.100 up
    route add -host 10.0.0.100 dev lo:0
root@ns2:/etc/keepalived# ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet 10.0.0.100/32 brd 10.0.0.100 scope global lo:0
       valid_lft forever preferred_lft forever
```
Thanks!!! One of those fix it when you ask it questions.
